### PR TITLE
feat:  implementation of dynamicAuthorization.deleteGroup

### PR DESCRIPTION
**Pull request by otlci**
Issue #, if available: MAFoundation-552

Description of changes:

Checklist:

* [x] Have you successfully deployed to an AWS account with your changes?
* [x] Have you written new tests for your core changes, as applicable?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: workbench-core/authorization/src/dynamicAuthorization/dynamicAuthorizationService.ts [hunk 1]
```diff
@@ -161,9 +163,63 @@ export class DynamicAuthorizationService {
...
+      } catch (error) {
+        metadata.statusCode = 400;
+        await this._auditService.write(metadata, error);
+        throw new RetryError(error.message);
+      }
+    }
+
+    try {
+      const response = await this._groupManagementPlugin.deleteGroup(deleteGroupRequest);
+
```
Comment (bhudnell) [inline workbench-core/authorization/src/dynamicAuthorization/dynamicAuthorizationService.ts:212]: We need to remove the `delete_pending` status once everything is completed gone right?

File: workbench-core/example/infrastructure/integration-tests/tests/dynamicAuthorization/group.test.ts [hunk 2]
```diff
@@ -289,4 +302,22 @@ describe('dynamic authorization group integration tests', () => {
...
+    it('deletes existing group', async () => {
+      const {
+        data: { groupId }
+      } = await adminSession.resources.groups.create();
+      const group = adminSession.resources.groups.group(groupId);
+
+      await group.delete();
+    });
+
+    it('returns a 404 error when trying to delete a group that doesn not exists', async () => {
```
Comment (bhudnell) [inline workbench-core/example/infrastructure/integration-tests/tests/dynamicAuthorization/group.test.ts:316]:
```suggestion
    it('returns a 404 error when trying to delete a group that does not exists', async () => {
```

File: workbench-core/example/infrastructure/integration-tests/tests/dynamicAuthorization/group.test.ts [hunk 3]
```diff
@@ -22,7 +22,6 @@ describe('dynamic authorization group integration tests', () => {
   let adminSession: ClientSession;
 
   beforeEach(() => {
-    expect.hasAssertions();
```
Comment (bhudnell) [inline workbench-core/example/infrastructure/integration-tests/tests/dynamicAuthorization/group.test.ts:25]: why remove `expect.hasAssertions()` from here?

File: workbench-core/authorization/src/dynamicAuthorization/dynamicAuthorizationService.ts (hunk 1 above)
Comment (otlci) [reply workbench-core/authorization/src/dynamicAuthorization/dynamicAuthorizationService.ts:212]: I'm not sure I understand, is status stored separately from group? I thought it's group attribute so as group is deleted it gets removed as well. @htztomic ?

File: workbench-core/example/infrastructure/integration-tests/support/resources/dynamicAuthorization/group.ts [hunk 4]
```diff
@@ -29,6 +31,11 @@ export default class Group extends Resource {
     return this._axiosInstance.get(`${this._api}/is-user-assigned/${userId}`);
   }
 
+  public async delete(): Promise<void> {
+    this._clientSession.removeCleanupTask(`group-${this.id}`);
```
Comment (bhudnell) [inline workbench-core/example/infrastructure/integration-tests/support/resources/dynamicAuthorization/group.ts:35]: if the delete API request isnt successful, we dont want to remove the cleanup task.

File: workbench-core/example/infrastructure/integration-tests/tests/dynamicAuthorization/group.test.ts (hunk 3 above)
Comment (otlci) [reply workbench-core/example/infrastructure/integration-tests/tests/dynamicAuthorization/group.test.ts:25]: First `deleteGroup` test doesn't have any expectations so it fails there. I added `hasAssertions` calls to other tests to maintain the logic.

File: workbench-core/example/app/src/routes/dynamicAuthorizationRoutes.ts [hunk 5]
```diff
@@ -106,6 +106,28 @@ export function setUpDynamicAuthorizationRoutes(router: Router, service: Dynamic
...
+          authenticatedUser: res.locals.user
+        });
+
+        res.status(204).send();
+      } catch (error) {
+        if (isGroupNotFoundError(error)) {
+          throw Boom.notFound(error.message);
+        }
+        if (isRetryError(error)) {
+          throw Boom.serverUnavailable('Request a retry');
```
Comment (bhudnell) [inline workbench-core/example/app/src/routes/dynamicAuthorizationRoutes.ts:124]: Can we put a more descriptive error message here?

File: workbench-core/example/app/src/routes/dynamicAuthorizationRoutes.ts [hunk 6]
```diff
@@ -106,6 +106,28 @@ export function setUpDynamicAuthorizationRoutes(router: Router, service: Dynamic
...
+  router.delete(
+    '/authorization/groups/:groupId',
+    wrapAsync(async (req: Request, res: Response) => {
+      try {
+        await service.deleteGroup({
+          groupId: req.params.groupId,
+          authenticatedUser: res.locals.user
+        });
+
+        res.status(204).send();
```
Comment (bhudnell) [inline workbench-core/example/app/src/routes/dynamicAuthorizationRoutes.ts:118]: lets return the `data` response object that includes the deleted group.

File: workbench-core/example/app/src/configs/staticRouteConfig.ts [hunk 7]
```diff
@@ -223,6 +223,14 @@ export const routesMap: RoutesMap = {
       }
     ]
   },
+  [`/authorization/groups/${groupIDRegExpAsString}`]: {
```
Comment (bhudnell) [inline workbench-core/example/app/src/configs/staticRouteConfig.ts:226]: do all the integ tests pass? We ran into an issue with the group RegExp being the last part of a route before.

File: workbench-core/example/app/src/routes/dynamicAuthorizationRoutes.ts (hunk 6 above)
Comment (otlci) [reply workbench-core/example/app/src/routes/dynamicAuthorizationRoutes.ts:118]: Isn't that redundant? `groupId` is already in URL path and standard DELETE verb usually returns 204 if resource was successfully deleted.

File: workbench-core/authorization/src/dynamicAuthorization/wbcGroupManagementPlugin.test.ts [hunk 8]
```diff
@@ -158,10 +159,27 @@ describe('WBCGroupManagemntPlugin', () => {
...
+    });
+
+    test.each([
+      [IdpUnavailableError, new IdpUnavailableError('test error')],
+      [PluginConfigurationError, new PluginConfigurationError('test error')],
+      [GroupNotFoundError, new RoleNotFoundError('test error')],
+      [TooManyRequestsError, new TooManyRequestsError('test error')],
+      [Error, new Error('test error')]
+    ])('throws exception %s when UserManagementService throws exception %s', async (expected, error) => {
+      wbcGroupManagementPlugin.getGroupStatus = jest.fn().mockResolvedValue({
```
Comment (bhudnell) [inline workbench-core/authorization/src/dynamicAuthorization/wbcGroupManagementPlugin.test.ts:175]: why are we mocking `getGroupStatus`?

File: workbench-core/example/app/src/configs/staticRouteConfig.ts (hunk 7 above)
Comment (otlci) [reply workbench-core/example/app/src/configs/staticRouteConfig.ts:226]: I had to shuffle them around so they are evaluated in correct order but yes, integration tests are passing.

File: workbench-core/authorization/src/dynamicAuthorization/wbcGroupManagementPlugin.test.ts (hunk 8 above)
Comment (otlci) [reply workbench-core/authorization/src/dynamicAuthorization/wbcGroupManagementPlugin.test.ts:175]: Thanks, good catch. Removed.

File: workbench-core/example/app/src/routes/dynamicAuthorizationRoutes.ts (hunk 5 above)
Comment (otlci) [reply workbench-core/example/app/src/routes/dynamicAuthorizationRoutes.ts:124]: Updated. I'm not 100% sure about messaging so if there are better suggestions I can update it to something else.

File: workbench-core/example/infrastructure/integration-tests/tests/dynamicAuthorization/group.test.ts (hunk 3 above)
Comment (bhudnell) [reply workbench-core/example/infrastructure/integration-tests/tests/dynamicAuthorization/group.test.ts:25]: every test should have at least one `expect`, even if it only checks that the function doesnt throw.

File: workbench-core/example/app/src/routes/dynamicAuthorizationRoutes.ts (hunk 6 above)
Comment (bhudnell) [reply workbench-core/example/app/src/routes/dynamicAuthorizationRoutes.ts:118]: its useful to return the deleted resource so the caller can use it if need be, it also matches or response from the service for consistency

**Pull request closed without merging**